torch/nn/modules/module.py

```diff
@@ -254,28 +268,35 @@ def register_forward_hook(self, hook):
         return handle
 
     def __call__(self, *input, **kwargs):
-        for hook in self._forward_pre_hooks.values():
-            hook(self, input)
-        result = self.forward(*input, **kwargs)
-        for hook in self._forward_hooks.values():
-            hook_result = hook(self, input, result)
-            if hook_result is not None:
-                raise RuntimeError(
-                    "forward hooks should never return any values, but '{}'"
-                    "didn't return None".format(hook))
-        if len(self._backward_hooks) > 0:
-            var = result
-            while not isinstance(var, Variable):
-                if isinstance(var, dict):
-                    var = next((v for v in var.values() if isinstance(v, Variable)))
-                else:
-                    var = var[0]
-            grad_fn = var.grad_fn
-            if grad_fn is not None:
-                for hook in self._backward_hooks.values():
-                    wrapper = functools.partial(hook, self)
-                    functools.update_wrapper(wrapper, hook)
-                    grad_fn.register_hook(wrapper)
+        tracing_state = torch.jit.get_tracing_state(_first_var(input, kwargs))
+        try:
+            if tracing_state:
+                tracing_state.push_scope('%s$%d' % (self.__class__.__name__, id(self)))
```

Comment:
Instead of naming the scope with the name of the `Module` class, I think it would be nicer if we assign the name of instance.
One way of achieving this would be to assign a `self._name` attribute into the class whenever we add a module. This way, the names of the scopes would be the same as the names that you give to the modules, and the same that you obtain when printing the network. Also, there would be no need of adding the `id(self)` to the scope, as the names would be unique.

What do you think?

Reply:
Yes, it would be good if the computed name here corresponds with however we identify multiple copies of a module inside a parent module.

Reply:
I thought about it, it would be good to assign a `_name` property to a specific instance of a module at the constructor level.
If the user doesn't provide the instance name, though, we have two options:
- use the `id`, quick but not that readable
- assign the name using a self-incrementing counter, as a class-level member variable; the gotcha here is what happens when one loads a serialized model
What do you think?

Reply:
There is always a name, because we store the modules in an `OrderedDict`, so that should be fine I think?
Let me know if this isn't clear, I can try writing a snippet.

Reply:
Ah-ha, you're right!

Reply:
One possibility for capturing the name of a module is to add it to the module in both [`__setattr__`](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L323-L328) and [`add_module`](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L111-L121). 
So you'd do something like (in `add_module`):
```python
module._name = name
self._modules[name] = module
```
Maybe it's not necessary to add a new attribute to the module, and we can use the names in `self._modules` directly.

On an unrelated note, it's a pity that `__setattr__` doesn't call `add_module` internally I think, but well, that's for another discussion.

Reply:
Done!